openwisp_ipam/api/utils.py

```diff
@@ -0,0 +1,85 @@
+from django.core.exceptions import ValidationError
+from rest_framework.exceptions import NotFound, PermissionDenied
+
+
+class FilterByOrganization:
+    def get_queryset(self):
+        qs = super().get_queryset()
+        # superuser has access to every organization
+        if self.request.user.is_superuser:
+            return qs
+        # non superuser has access only to some organizations
+        return self.get_organization_queryset(qs)
+
+    def get_organization_queryset(self):
+        raise NotImplementedError()
+
+
+class FilterByOrganizationManaged(FilterByOrganization):
+    """
+    Allows to filter only organizations which the current user manages
+    """
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=self.request.user.organizations_managed)
+
+
+class FilterByParent:
+    def get_queryset(self):
+        qs = super().get_queryset()
+        self.assert_parent_exists()
+        return qs
+
+    def assert_parent_exists(self):
+        parent_queryset = self.get_parent_queryset()
+        if not self.request.user.is_superuser:
+            parent_queryset = self.get_organization_queryset(parent_queryset)
+        try:
+            assert parent_queryset.exists()
+        except (AssertionError, ValidationError):
+            raise NotFound(detail='No relevant data found.')
+
+    def get_parent_queryset(self):
+        raise NotImplementedError()
+
+    def get_organization_queryset(self):
+        raise NotImplementedError()
+
+
+class FilterByParentManaged(FilterByParent):
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=self.request.user.organizations_managed)
+
+
+class AuthorizeCSVImport:
+    def post(self, request):
+        self.assert_organization_permissions(request)
+
+    def assert_organization_permissions(self, request):
+        if request.user.is_superuser:
+            return
+        organization = self.get_csv_organization()
+        if not organization:
+            # if organization in CSV doesn't exist, then check if
+            # user can create new organizations
+            create_org_permission = request.user.permissions.filter(
+                codename='add_organization'
+            )
+            if create_org_permission.exists():
+                return
+        else:
+            # else check if user has access to organization in CSV
+            if str(organization.pk) in self.get_user_organizations():
+                return
+        raise PermissionDenied()
+
+    def get_csv_organization(self):
+        raise NotImplementedError()
+
+    def get_user_organizations(self):
+        raise NotImplementedError()
+
+
+class AuthorizeCSVOrgManaged(AuthorizeCSVImport):
+    def get_user_organizations(self):
+        return self.request.user.organizations_managed
```

Comment:
@nemesisdesign I made this reusable class to authorize based on organization in a `.csv` file. Is there a scope for improvement here?